gax-httpjson/src/main/java/com/google/api/gax/httpjson/InstantiatingHttpJsonChannelProvider.java

```diff
@@ -193,30 +199,38 @@ public static Builder newBuilder() {
   }
 
   public static final class Builder {
-    private ExecutorProvider executorProvider;
+    private Executor executor;
     private HeaderProvider headerProvider;
     private String endpoint;
     private HttpTransport httpTransport;
 
     private Builder() {}
 
     private Builder(InstantiatingHttpJsonChannelProvider provider) {
-      this.executorProvider = provider.executorProvider;
+      this.executor = provider.executor;
       this.headerProvider = provider.headerProvider;
       this.endpoint = provider.endpoint;
       this.httpTransport = provider.httpTransport;
     }
 
+    /** @deprecated Please use {@link #setExecutor(Executor)} instead. */
+    @Deprecated
+    public Builder setExecutorProvider(ExecutorProvider executorProvider) {
+      Executor executor = null;
+      if (executorProvider != null) {
+        executor = executorProvider.getExecutor();
```

Comment:
For InstantiatingExecutorProvider `getExecutor()` creates a new instance each time (for each channel, i guess). Calling it preemptively and only once basically converts `InstantiatingExecutorProvider` into a `FixedExecutorProvider`. This seems dangerous  and may lead to extremely difficult to discover bugs. 

Given that I'm not sure that I understand why the decision was made to "migrate" to explicitly setting Executor (i.e. adding `setExecutor()` and deprecating `setExecutorProvider()`). I.e. maybe we can keep the `ExecutorProvider` thing as is, and the only thing which will be changed is the default behavior to not overriding grpc's executor if user hasn't touched default settings for it in his client/stub settings?

Reply:
The common code path is:
ClientContext.create()  -> transportProvider.withExecutor(executor) -> transportProvider.executorProvider = FixedExecutorProvider.create(executor)

So in most circumstances the behavior of channel sharing an executor is the same, which is good: you don't want a separate executor per channel.

The main behavior change  is when the caller does something like:
stubSettings.setTransportProvider(
   InstantiatingGrpcChannelProvider.newBuilder()
     .setExecutorProvider(
         InstantiatingExecutorProvider.newBuilder().build()
     )
)
This would create an executor per channel, which is different behavior from
stubSettings.setExecutorProvider(
   InstantiatingExecutorProvider.newBuilder().build()
)
Which would share an executor across all channels

99% of the time you don't want a separate executor per channel

This change makes this explicit and removes ambiguity